enforcer/src/keystate/keystate_import_cmd.c

```diff
@@ -0,0 +1,467 @@
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+    key = key_data_new_get_by_hsm_key_id(dbconn, hsmkeyid);
+    keydataid = key_data_id(key);
+
+    policy = policy_new(dbconn);
+    policy_get_by_id(policy, zone_policy_id(zone_new_get_by_name(dbconn, zonename)));
+
+    if (!(key_state = key_state_new(dbconn))
+                || key_state_set_key_data_id(key_state, keydataid)
+                || key_state_set_type(key_state, KEY_STATE_TYPE_DS)
+                || key_state_set_last_change (key_state, time)
+                || key_state_set_minimize(key_state, (key_data_minimize(key) >> 2) & 1)
+                || key_state_set_ttl (key_state, policy_parent_ds_ttl(policy))
+		|| key_state_set_state(key_state, keytype == 1 ? ksk_mapping[keystate][0] : zsk_mapping[keystate][0])
+                || key_state_create(key_state))
+    {
+        ods_log_error("[%s] key state creation for DS failed, database or memory error", module_str);
+        key_state_free(key_state);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
+
+    if (!(key_state = key_state_new(dbconn))
+                || key_state_set_key_data_id(key_state, keydataid)
+                || key_state_set_type(key_state, KEY_STATE_TYPE_DNSKEY)
+                || key_state_set_last_change (key_state, time)
+                || key_state_set_minimize(key_state, (key_data_minimize(key) >> 1) & 1)
+                || key_state_set_ttl (key_state, policy_keys_ttl(policy))
+		|| key_state_set_state(key_state, keytype == 1 ? ksk_mapping[keystate][1] : zsk_mapping[keystate][1])
+                || key_state_create(key_state))
+    {
+        ods_log_error("[%s] key state creation for DNSKEY failed, database or memory error", module_str);
+        key_state_free(key_state);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
+    if (!(key_state = key_state_new(dbconn))
+                || key_state_set_key_data_id(key_state, keydataid)
+                || key_state_set_type(key_state, KEY_STATE_TYPE_RRSIGDNSKEY)
+                || key_state_set_last_change (key_state, time)
+                || key_state_set_ttl (key_state, policy_keys_ttl(policy))
+                || key_state_set_state(key_state, keytype == 1 ? ksk_mapping[keystate][2] : zsk_mapping[keystate][2])
+                || key_state_create(key_state))
+    {
+        ods_log_error("[%s] key state creation for RRSIGDNSKEY failed, database or memory error", module_str);
+        key_state_free(key_state);
+        hsm_destroy_context(hsm_ctx);
+        return -1;
+    }
+
+    ttl = max(min(policy_zone_soa_ttl(policy), policy_zone_soa_minimum(policy)),
+            (policy_denial_type(policy) == POLICY_DENIAL_TYPE_NSEC3
+                ? ( policy_denial_ttl(policy) > policy_signatures_max_zone_ttl(policy)
+                    ? policy_denial_ttl(policy)
+                    : policy_signatures_max_zone_ttl(policy))
+                : policy_signatures_max_zone_ttl(policy)));
+
```

Comment:
This is horrible. (I see you copied it from enforcer.c, not blaming you.) We should consider rewriting this to something readable...